File: State_Pattern/sp.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>
#include <climits>
#include <cstdlib>

class GumballMachine;

class State {
   public:
    virtual void insertQuater() = 0;
    virtual void ejectQuater() = 0;
    virtual void turnCrank() = 0;
    virtual void dispense() = 0;
};

class SoldoutState : public State {
    std::shared_ptr<GumballMachine> machine;

public:
    explicit SoldoutState(std::weak_ptr<GumballMachine> machine)
        : machine(machine.lock()) {}

    void insertQuater() override { std::cout << "You can't insert a quarter, the machine is sold out\n"; }

    void ejectQuater() override { std::cout << "You can't eject, you haven't inserted a quarter yet\n"; }
    void turnCrank() override { std::cout << "You turned, but there are no gumballs\n"; }
    void dispense() override { std::cout << "No gumball dispensed\n"; }
};

class NoQuaterState : public State {
    std::shared_ptr<GumballMachine> machine;

   public:
    explicit NoQuaterState(std::weak_ptr<GumballMachine> machine)
        : machine(machine.lock()) {}

    void insertQuater() override {
        std::cout << "You inserted a quater!\n";
        this->machine->setState(this->machine->hasQuarterState);
    }

    void ejectQuater() override { std::cout << "You haven't inserted a quarter\n"; }
    void turnCrank() override { std::cout << "You turned, but there's no quarter\n"; }
    void dispense() override { std::cout << "You need to pay first\n"; }
};

class HasQuaterState : public State {

    std::shared_ptr<GumballMachine> machine;

   public:
    explicit HasQuaterState(std::weak_ptr<GumballMachine> machine)
        : machine(machine.lock()) {}

    void ejectQuater() override {
        std::cout << "Quarter returned\n";
        this->machine->setState(this->machine->noQuarterState);
    }

    void turnCrank() override {
        auto randomNum = []() {return (std::rand()/RAND_MAX)*10;};
        std::cout << "You turned...\n";
        if (randomNum() and this->machine->count > 1)
            this->machine->setState(this->machine->winnerState);
        else 
            this->machine->setState(this->machine->soldState);
    }

    void insertQuater() override { std::cout << "You can't insert another quarter!\n"; }
    void dispense() override { std::cout << "No gumball dispensed\n"; }
};

class SoldState : public State {
    std::shared_ptr<GumballMachine> machine;

   public:
    explicit SoldState(std::weak_ptr<GumballMachine> machine)
        : machine(machine.lock()) {}

    void insertQuater() override { std::cout << "Please wait, we're already giving you a gumball\n"; }
    void ejectQuater() override { std::cout << "Sorry, you already turned the crank\n"; }
    void turnCrank() override { std::cout << "Turning twice doesn't get you another gumball\n"; }

    void dispense() override {
        this->machine->releaseBall();
        if (this->machine->count > 0) {
            this->machine->setState(this->machine->noQuarterState);
        } else {
            std::cout << "Oops, out of gumballs!\n";
            this->machine->setState(this->machine->soldOutState);
        }
    }
};

class WinnerState : public State {
    std::shared_ptr<GumballMachine> machine;

   public:
    explicit WinnerState(std::weak_ptr<GumballMachine> machine)
        : machine(machine.lock()) {}

    void insertQuater() override { std::cout << "Please wait, we're already giving you a gumball\n"; }
    void ejectQuater() override { std::cout << "Sorry, you already turned the crank\n"; }
    void turnCrank() override { std::cout << "Turning twice doesn't get you another gumball\n"; }

    void dispense() override {
        std::cout << "YOU'RE A WINNER! You get two gumballs for your quarter\n";
        this->machine->releaseBall();
        if (this->machine->count == 0) {
            this->machine->setState(this->machine->soldOutState);
        } else {
            this->machine->releaseBall();
            if (this->machine->count > 0) {
                this->machine->setState(this->machine->noQuarterState);
            } else {
                std::cout << "Oops, out of gumballs!\n";
                this->machine->setState(this->machine->soldOutState);
            }
        }
    }
};

class GumballMachine {
   public:
    std::shared_ptr<State> soldOutState;
    std::shared_ptr<State> noQuarterState;
    std::shared_ptr<State> hasQuarterState;
    std::shared_ptr<State> soldState;
    std::shared_ptr<State> winnerState;

    std::shared_ptr<State> currState;

    int count = 0;

    explicit GumballMachine(const int count) {
        this->soldOutState = std::make_shared<SoldoutState>(*this);
        this->noQuarterState = std::make_shared<NoQuaterState>(*this);
        this->hasQuarterState = std::make_shared<HasQuaterState>(*this);
        this->soldState = std::make_shared<SoldState>(*this);

        this->count = count;
        if (count > 0)
            this->currState = this->hasQuarterState;
    }

    void insertQuater() { this->currState->insertQuater(); }
    void ejectQuater() { this->currState->ejectQuater(); }

    void turnCrank() {
        this->currState->turnCrank();
        this->currState->dispense();
    }

    void setState(std::weak_ptr<State> state) { this->currState = state.lock(); }

    void releaseBall() {
        std::cout << "A gumball comes rolling out the slot...\n";
        if (this->count != 0)
            this->count--;
    }
};

int main() {
    /*
            The code will work when gumballmachine class is declred on top and its functions are defined before main
    */
    auto machine = std::make_shared<GumballMachine>(5);

    machine->insertQuater();
    machine->turnCrank();

    machine->insertQuater();
    machine->ejectQuater();
    machine->turnCrank();

    machine->insertQuater();
    machine->turnCrank();
    machine->insertQuater();
    machine->turnCrank();
    machine->ejectQuater();

    machine->insertQuater();
    machine->insertQuater();
    machine->turnCrank();
    machine->insertQuater();
    machine->turnCrank();
    machine->insertQuater();
    machine->turnCrank();
}

File: State_Pattern/sp.py
from abc import ABC, abstractmethod
from random import randint

class State:
    @abstractmethod
    def insertQuater(self) -> None: pass

    @abstractmethod
    def ejectQuater(self) -> None: pass

    @abstractmethod
    def turnCrank(self) -> None: pass

    @abstractmethod
    def dispense(self) -> None: pass


class soldoutState(State):
    def __init__(self, machine: 'GumballMachine') -> None:
        self.machine: GumballMachine = machine

    def insertQuater(self) -> None: print("You can't insert a quarter, the machine is sold out")
    def ejectQuater(self) -> None: print("You can't eject, you haven't inserted a quarter yet")
    def turnCrank(self) -> None: print("You turned, but there are no gumballs")
    def dispense(self) -> None: print("No gumball dispensed")

    
class NoQuaterState(State):
    def __init__(self, machine: 'GumballMachine') -> None:
        self.machine: GumballMachine = machine

    def insertQuater(self) -> None: 
        print("You inserted a quater!")
        self.machine.setState(self.machine.hasQuarterState)
        
    def ejectQuater(self) -> None: print("You haven't inserted a quarter")
    def turnCrank(self) -> None: print("You turned, but there's no quarter")
    def dispense(self) -> None: print("You need to pay first")


class WinnerState(State):
    def __init__(self, machine: 'GumballMachine') -> None:
        self.machine: GumballMachine = machine

    def insertQuater(self) -> None: print("Please wait, we're already giving you a gumball")
    def ejectQuater(self) -> None: print("Sorry, you already turned the crank")
    def turnCrank(self) -> None: print("Turning twice doesn't get you another gumball")
    def dispense(self) -> None: 
        print("YOU'RE A WINNER! You get two gumballs for your quarter")
        self.machine.releaseBall()
        if self.machine.count == 0:
            self.machine.setState(self.machine.soldOutState)
        else:
            self.machine.releaseBall()
            if self.machine.count > 0:
                self.machine.setState(self.machine.noQuarterState)
            else:
                print("Oops, out of gumballs!")
                self.machine.setState(self.machine.soldOutState)
    
    
class HasQuaterState(State):
    def __init__(self, machine: 'GumballMachine') -> None:
        self.machine: GumballMachine = machine

    def insertQuater(self) -> None: print("You can't insert another quarter!")
    
    def ejectQuater(self) -> None: 
        print("Quarter returned")
        self.machine.setState(self.machine.noQuarterState)
        
    def turnCrank(self) -> None: 
        if randint(1, 10) == 5 and self.machine.count > 1:
            self.machine.setState(self.machine.winnerState)
        else:
            self.machine.setState(self.machine.soldState)
        
    def dispense(self) -> None: print("No gumball dispensed")

    
class SoldState(State):
    def __init__(self, machine: 'GumballMachine') -> None:
        self.machine: GumballMachine = machine

    def insertQuater(self) -> None: print("Please wait, we're already giving you a gumball")
    def ejectQuater(self) -> None: print("Sorry, you already turned the crank")
    def turnCrank(self) -> None: print("Turning twice doesn't get you another gumball")
    
    def dispense(self) -> None: 
        self.machine.releaseBall()
        if self.machine.count > 0:
            self.machine.setState(self.machine.noQuarterState)
        else:
            print("Oops, out of gumballs!")
            self.machine.setState(self.machine.soldOutState)
    

class GumballMachine:
    def __init__(self, count: int) -> None:
        self.soldOutState: State = SoldState(self)
        self.noQuarterState: State = NoQuaterState(self)
        self.hasQuarterState: State = HasQuaterState(self)
        self.soldState: State = SoldState(self)
        self.winnerState: State = WinnerState(self)
        
        self.count: int = count
        if self.count > 0:
            self.currState: State = self.hasQuarterState

    def insertQuater(self) -> None: self.currState.insertQuater()
    def ejectQuater(self) -> None: self.currState.ejectQuater()

    def turnCrank(self) -> None:
        self.currState.turnCrank()
        self.currState.dispense()

    def setState(self, state: State) -> None: self.currState = state

    def releaseBall(self) -> None:
        print("A gumball comes rolling out the slot...")
        if self.count != 0: self.count -=1


def main():
    machine = GumballMachine(5)
    machine.insertQuater()
    machine.turnCrank()

    machine.insertQuater()
    machine.ejectQuater()
    machine.turnCrank()

    machine.insertQuater()
    machine.turnCrank()
    machine.insertQuater()
    machine.turnCrank()


if __name__ == '__main__':
    main()